python/rmm/_lib/logger.pyx

```diff
@@ -0,0 +1,185 @@
+    cdef enum level_enum:
+        trace = 0
+        debug = 1
+        info = 2
+        warn = 3
+        err = 4
+        critical = 5
+        off = 6
+
+
+cdef extern from "spdlog/spdlog.h" namespace "spdlog" nogil:
+    cdef cppclass spdlog_logger "spdlog::logger":
+        spdlog_logger() except +
+        void set_level(level_enum log_level) except +
+        level_enum level() except +
+        void flush() except +
+        void flush_on(level_enum log_level) except +
+        level_enum flush_level() except +
+
+
+cdef extern from "rmm/logger.hpp" namespace "rmm" nogil:
+    cdef spdlog_logger& logger() except +
+
+
+logging_levels = [
+    "trace",
+    "debug",
+    "info",
+    "warn",
+    "err",
+    "critical",
+    "off",
+]
+
+
+def _normalize_logging_level(level):
+    if isinstance(level, str):
+        try:
+            level = logging_levels.index(level.lower())
+        except ValueError:
+            raise ValueError(
+                f"Invalid logging level '{level}'. Valid levels are "
+                f"{logging_levels}."
+            )
+    elif not isinstance(level, int):
+        raise TypeError(
+            f"Logging level must be an integer or string, not {type(level)}"
+        )
+    else:
+        if level < 0 or level >= len(logging_levels):
+            raise ValueError(
+                f"Logging level must be between 0 and {len(logging_levels)-1}"
+            )
+
+    return level
+
+
+def set_logging_level(level):
+    """
+    Set the debug logging level for the RMM library.
+
+    Parameters
+    ----------
+    level : int or str
+        The debug logging level. Valid string names are (in decreasing order
+        of verbosity) "trace", "debug", "info", "warn", "err", "critical",
+        and "off", corresponding to valid integer levels 0 through 6. Default
+        is 2 (info).
+
+    See Also
+    --------
+    get_logging_level : Get the current debug logging level for the RMM
+
+    Examples
+    --------
+    >>> import rmm
+    >>> rmm.set_logging_level("debug") # set logging level to debug
+    >>> rmm.set_logging_level(3) # set logging level to warn
+    """
+    logger().set_level(_normalize_logging_level(level))
+
+
+def get_logging_level():
+    """
+    Get the current debug logging level for the RMM library.
+
+    Returns
+    -------
+    level : int
+    The current debug logging level. Valid values are 0 through 6 inclusive,
+    where 0 is the most verbose and 6 is the least verbose. Default is 2
+    (info).
+
+    See Also
+    --------
+    set_logging_level : Set the debug logging level for the RMM library.
+
+    Examples
+    --------
+    >>> import rmm
+    >>> rmm.get_logging_level() # get current logging level
+    2
+    """
+    return logger().level()
+
+
+def flush_logger():
+    """
+    Flush the RMM logger.
+
+    See Also
+    --------
+    set_logging_level : Set the debug logging level for the RMM library.
```

Comment:
This seems like a strange "See Also", I don't see a connection between flushing and setting the level (or at least it seems no better than linking all the logger functions).

Same question for `set_flush_level`'s See Also `get_logging_level` and `get_flush_level`'s See Also `set_logging_level`.

Reply:
Yeah, GPT/copilot generated these based on my first examples. Should have notice this.

Reply:
Improved